### PR TITLE
Implement Deidentification Endpoint

### PR DESCRIPTION
**Pull request by cascadianblue**
@tschaffter : I'm creating this PR before I have completely finished the "annotation type" de-identification method, because I want you to get a chance to look at how I'm doing things sooner rather than later. I.e. it would be great if you could look at this, but I am planning on adding some more commits before attempting to merge.

One thing that I know has to be done differently is the unit testing (see: `server/openapi_server/test/test_deidentified_notes_controller.py`). This unit test file was originally auto-generated and was supposed to integrate with flask, but I wasn't immediately sure how to get that to work given that we use an entire dockerized stack, so I just changed them to call a local server using blackbox HTTP calls via `requests.post()`.

Main question is: should I try to get the unit tests to work in the particular flask-integrated way they were intended to run, or should I just yank them out of the `server` directory and run them as a separate concern? Also, do we want the tests to be part of CI?

### CODE REVIEW

**Comment by cascadianblue** [comment]
@tschaffter this is ready for your review

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -32,6 +32,14 @@ jobs:
     - uses: actions/checkout@v2
     - name: Validate docker-compose.yml
       run: docker-compose -f docker-compose.yml config >/dev/null
+    - name: Set up Python 3.8
```
Comment (tschaffter) [inline .github/workflows/ci.yml:35]: I'm using `tox` in other repositories to run the test suite. Will be address in https://github.com/nlpsandbox/phi-deidentifier/issues/22.

File: server/openapi_server/controllers/deidentified_notes_controller.py [hunk 2]
```diff
@@ -1,72 +1,70 @@
...
+                    deid_config.annotation_types
+                )
+            elif deid_config.deidentification_strategy.annotation_type_config is not None:
+                deidentified_note, deidentified_annotations = apply_annotation_type(
+                    deidentified_note,
+                    deidentified_annotations,
+                    deid_config.annotation_types
+                )
+            else:
+                return "No supported de-identification method supported in request: '%s'" % (str(deid_config.to_dict()),), 400
```
Comment (tschaffter) [inline server/openapi_server/controllers/deidentified_notes_controller.py:55]: Need to return an `Error` object. Here is the list of https://github.com/nlpsandbox/nlpsandbox-schemas/tree/develop/openapi/commons/components/responses. If none are suitable, please open a ticket/PR to add a new response that will then be referenced by the deidentifier OpenAPI spec.

File: server/openapi_server/controllers/deidentified_notes_controller.py [hunk 3]
```diff
@@ -1,72 +1,70 @@
...
+                text_person_name_annotations=annotations['text_person_name'],
+                text_physical_address_annotations=annotations['text_physical_address']
+            ),
+            deidentified_annotations=Annotation(
+                text_date_annotations=deidentified_annotations['text_date'],
+                text_person_name_annotations=deidentified_annotations['text_person_name'],
+                text_physical_address_annotations=deidentified_annotations['text_physical_address']
+            )
+        )
+        return deidentify_response, 201
```
Comment (tschaffter) [inline server/openapi_server/controllers/deidentified_notes_controller.py:70]: Return status code 200 (see https://github.com/nlpsandbox/nlpsandbox-schemas/issues/86#issue-747886275)

File: server/openapi_server/controllers/service_controller.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+def service():  # noqa: E501
+    """Get service information
+
+    Get information about the service # noqa: E501
+
+
+    :rtype: Service
+    """
+    return 'do some magic!'
```
Comment (tschaffter) [inline server/openapi_server/controllers/service_controller.py:17]: Will be implemented in https://github.com/nlpsandbox/phi-deidentifier/issues/23

File: server/openapi_server/models/__init__.py [hunk 5]
```diff
@@ -3,8 +3,18 @@
 # flake8: noqa
 from __future__ import absolute_import
 # import models into model package
-from openapi_server.models.entity import Entity
+from openapi_server.models.annotation import Annotation
+from openapi_server.models.annotation_source import AnnotationSource
+from openapi_server.models.deidentification_config import DeidentificationConfig
+from openapi_server.models.deidentification_config_deidentification_strategy import DeidentificationConfigDeidentificationStrategy
+from openapi_server.models.deidentification_config_deidentification_strategy_date_offset_config import DeidentificationConfigDeidentificationStrategyDateOffsetConfig
```
Comment (tschaffter) [inline server/openapi_server/models/__init__.py:10]: These variable names are quite long. We could have another look at the API specification of the deidentifier to see if we can simplify these names. Will be done in https://github.com/nlpsandbox/nlpsandbox-schemas/issues/123

File: server/openapi_server/test/test_deidentified_notes_controller.py [hunk 6]
```diff
@@ -2,36 +2,212 @@
...
+        }
+
+        response = self.client.open(
+            DEIDENTIFIER_ENDPOINT_URL,
+            method='POST',
+            headers={'Accept': 'application/json'},
+            data=json.dumps(redact_request),
+            content_type='application/json'
+        )
+        self.assertStatus(response, 201, 'Response body is : ' + response.data.decode('utf-8'))
```
Comment (tschaffter) [inline server/openapi_server/test/test_deidentified_notes_controller.py:78]: Must be 200 once changed in the controller. First make sure the deidentifier specification states 200, if not open a PR to implement this schema change.

File: server/openapi_server/utils/annotator_client.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import requests
```
Comment (tschaffter) [inline server/openapi_server/utils/annotator_client.py:1]: I'm not a huge fan of the utils class as it's often used as the default place to store code that one doesn't know where to place. We can keep as it is for now and revisit later.

**Review by tschaffter — APPROVED** [review]

File: server/openapi_server/controllers/deidentified_notes_controller.py [hunk 8]
```diff
@@ -2,71 +2,163 @@
...
+                deidentified_note, deidentified_annotations = apply_redaction(
+                    deidentified_note,
+                    deidentified_annotations,
+                    deid_config['annotationTypes']
+                )
+            else:
+                # Handle other deid methods in elif's
+                pass
+
+        return {
```
[the remark this answers is not in the document]
Comment (cascadianblue) [reply server/openapi_server/controllers/deidentified_notes_controller.py]: @tschaffter  hmmm, you seem to be looking at an old version of this file? [This](https://github.com/nlpsandbox/phi-deidentifier/blob/develop/server/openapi_server/controllers/deidentified_notes_controller.py#L70) is what it looks like now

**Pull request merged**